Errors when deferring assertions

When attempting to `defer` an assert statement in certain scenarios, I get the following error message:

```
failed to find an expression on line XX in /path/to/my/file_test.go
```

The smallest example test I've been able to reproduce the error in is this:
```go
func TestDefer(t *testing.T) {
	defer assert.Check(t, false)
}
```

When asserting `true` instead of `false`, the test is successful, so my guess is the bug is limited to the error reporting logic.

This may be a separate issue, but when wrapping the assertion in an anonymous function call, the error becomes a panic:
```go
func TestDefer(t *testing.T) {
	defer func() { assert.Check(t, false) }()
}
```

With the following error, again only occurring on failed assertions:
```
--- FAIL: TestDefer (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range
```

--- a/assert/result.go
+++ b/assert/result.go
@@ -70,7 +70,6 @@ func filterPrintableExpr(args []ast.Expr) []ast.Expr {
 			result[i] = starExpr.X
 			continue
 		}
-		result[i] = nil
 	}
 	return result
 }

--- a/dobi.yaml
+++ b/dobi.yaml
@@ -33,7 +33,9 @@ job=watch:
   interactive: true
   command: |
       filewatcher -x vendor gotestsum -- -test.timeout 10s
-  env: ['GOTESTSUM_FORMAT=short-verbose']
+  env:
+    - 'GOTESTSUM_FORMAT=short-verbose'
+    - 'GOTESTTOOLS_DEBUG={env.GOTESTTOOLS_DEBUG:}'
 
 job=test-unit:
   use: builder

--- a/internal/source/defers.go
+++ b/internal/source/defers.go
@@ -1,0 +1,53 @@
+package source
+
+import (
+	"go/ast"
+	"go/token"
+
+	"github.com/pkg/errors"
+)
+
+func scanToDeferLine(fileset *token.FileSet, node ast.Node, lineNum int) ast.Node {
+	var matchedNode ast.Node
+	ast.Inspect(node, func(node ast.Node) bool {
+		switch {
+		case node == nil || matchedNode != nil:
+			return false
+		case fileset.Position(node.End()).Line == lineNum:
+			if funcLit, ok := node.(*ast.FuncLit); ok {
+				matchedNode = funcLit
+				return false
+			}
+		}
+		return true
+	})
+	debug("defer line node: %s", debugFormatNode{matchedNode})
+	return matchedNode
+}
+
+func guessDefer(node ast.Node) (ast.Node, error) {
+	defers := collectDefers(node)
+	switch len(defers) {
+	case 0:
+		return nil, errors.New("failed to expression in defer")
+	case 1:
+		return defers[0].Call, nil
+	default:
+		return nil, errors.Errorf(
+			"ambiguous call expression: multiple (%d) defers in call block",
+			len(defers))
+	}
+}
+
+func collectDefers(node ast.Node) []*ast.DeferStmt {
+	var defers []*ast.DeferStmt
+	ast.Inspect(node, func(node ast.Node) bool {
+		if d, ok := node.(*ast.DeferStmt); ok {
+			defers = append(defers, d)
+			debug("defer: %s", debugFormatNode{d})
+			return false
+		}
+		return true
+	})
+	return defers
+}

--- a/internal/source/source.go
+++ b/internal/source/source.go
@@ -97,6 +97,7 @@ func getCallExprArgs(node ast.Node) ([]ast.Expr, error) {
 	if visitor.expr == nil {
 		return nil, errors.New("failed to find call expression")
 	}
+	debug("using (%T): %s", visitor.expr, debugFormatNode{visitor.expr})
 	return visitor.expr.Args, nil
 }
 
@@ -110,8 +111,12 @@ func (v *callExprVisitor) Visit(node ast.Node) ast.Visitor {
 	}
 	debug("visit (%T): %s", node, debugFormatNode{node})
 
-	if callExpr, ok := node.(*ast.CallExpr); ok {
-		v.expr = callExpr
+	switch typed := node.(type) {
+	case *ast.DeferStmt:
+		ast.Walk(v, typed.Call.Fun)
+		return nil
+	case *ast.CallExpr:
+		v.expr = typed
 		return nil
 	}
 	return v
@@ -142,7 +147,7 @@ func CallExprArgs(stackIndex int) ([]ast.Expr, error) {
 	return getCallExprArgs(node)
 }
 
-var debugEnabled = os.Getenv("GOTESTYOURSELF_DEBUG") != ""
+var debugEnabled = os.Getenv("GOTESTTOOLS_DEBUG") != ""
 
 func debug(format string, args ...interface{}) {
 	if debugEnabled {

--- a/internal/source/source.go
+++ b/internal/source/source.go
@@ -24,109 +24,10 @@ func FormattedCallExprArg(stackIndex int, argPos int) (string, error) {
 	if err != nil {
 		return "", err
 	}
+	if argPos >= len(args) {
+		return "", errors.New("no arg")
+	}
 	return FormatNode(args[argPos])
-}
-
-func getNodeAtLine(filename string, lineNum int) (ast.Node, error) {
-	fileset := token.NewFileSet()
-	astFile, err := parser.ParseFile(fileset, filename, nil, parser.AllErrors)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to parse source file: %s", filename)
-	}
-
-	node := scanToLine(fileset, astFile, lineNum)
-	if node == nil {
-		return nil, errors.Errorf(
-			"failed to find an expression on line %d in %s", lineNum, filename)
-	}
-	return node, nil
-}
-
-func scanToLine(fileset *token.FileSet, node ast.Node, lineNum int) ast.Node {
-	v := &scanToLineVisitor{lineNum: lineNum, fileset: fileset}
-	ast.Walk(v, node)
-	return v.matchedNode
-}
-
-type scanToLineVisitor struct {
-	lineNum     int
-	matchedNode ast.Node
-	fileset     *token.FileSet
-}
-
-func (v *scanToLineVisitor) Visit(node ast.Node) ast.Visitor {
-	if node == nil || v.matchedNode != nil {
-		return nil
-	}
-	if v.nodePosition(node).Line == v.lineNum {
-		v.matchedNode = node
-		return nil
-	}
-	return v
-}
-
-// In golang 1.9 the line number changed from being the line where the statement
-// ended to the line where the statement began.
-func (v *scanToLineVisitor) nodePosition(node ast.Node) token.Position {
-	if goVersionBefore19 {
-		return v.fileset.Position(node.End())
-	}
-	return v.fileset.Position(node.Pos())
-}
-
-var goVersionBefore19 = isGOVersionBefore19()
-
-func isGOVersionBefore19() bool {
-	version := runtime.Version()
-	// not a release version
-	if !strings.HasPrefix(version, "go") {
-		return false
-	}
-	version = strings.TrimPrefix(version, "go")
-	parts := strings.Split(version, ".")
-	if len(parts) < 2 {
-		return false
-	}
-	minor, err := strconv.ParseInt(parts[1], 10, 32)
-	return err == nil && parts[0] == "1" && minor < 9
-}
-
-func getCallExprArgs(node ast.Node) ([]ast.Expr, error) {
-	visitor := &callExprVisitor{}
-	ast.Walk(visitor, node)
-	if visitor.expr == nil {
-		return nil, errors.New("failed to find call expression")
-	}
-	debug("using (%T): %s", visitor.expr, debugFormatNode{visitor.expr})
-	return visitor.expr.Args, nil
-}
-
-type callExprVisitor struct {
-	expr *ast.CallExpr
-}
-
-func (v *callExprVisitor) Visit(node ast.Node) ast.Visitor {
-	if v.expr != nil || node == nil {
-		return nil
-	}
-	debug("visit (%T): %s", node, debugFormatNode{node})
-
-	switch typed := node.(type) {
-	case *ast.DeferStmt:
-		ast.Walk(v, typed.Call.Fun)
-		return nil
-	case *ast.CallExpr:
-		v.expr = typed
-		return nil
-	}
-	return v
-}
-
-// FormatNode using go/format.Node and return the result as a string
-func FormatNode(node ast.Node) (string, error) {
-	buf := new(bytes.Buffer)
-	err := format.Node(buf, token.NewFileSet(), node)
-	return buf.String(), err
 }
 
 // CallExprArgs returns the ast.Expr slice for the args of an ast.CallExpr at
@@ -145,6 +46,98 @@ func CallExprArgs(stackIndex int) ([]ast.Expr, error) {
 	debug("found node (%T): %s", node, debugFormatNode{node})
 
 	return getCallExprArgs(node)
+}
+
+func getNodeAtLine(filename string, lineNum int) (ast.Node, error) {
+	fileset := token.NewFileSet()
+	astFile, err := parser.ParseFile(fileset, filename, nil, parser.AllErrors)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse source file: %s", filename)
+	}
+
+	node := scanToLine(fileset, astFile, lineNum)
+	if node == nil {
+		return nil, errors.Errorf(
+			"failed to find an expression on line %d in %s", lineNum, filename)
+	}
+	return node, nil
+}
+
+func scanToLine(fileset *token.FileSet, node ast.Node, lineNum int) ast.Node {
+	var matchedNode ast.Node
+	ast.Inspect(node, func(node ast.Node) bool {
+		if node == nil || matchedNode != nil {
+			return false
+		}
+		if nodePosition(fileset, node).Line == lineNum {
+			matchedNode = node
+			return false
+		}
+		return true
+	})
+	return matchedNode
+}
+
+// In golang 1.9 the line number changed from being the line where the statement
+// ended to the line where the statement began.
+func nodePosition(fileset *token.FileSet, node ast.Node) token.Position {
+	if goVersionBefore19 {
+		return fileset.Position(node.End())
+	}
+	return fileset.Position(node.Pos())
+}
+
+var goVersionBefore19 = func() bool {
+	version := runtime.Version()
+	// not a release version
+	if !strings.HasPrefix(version, "go") {
+		return false
+	}
+	version = strings.TrimPrefix(version, "go")
+	parts := strings.Split(version, ".")
+	if len(parts) < 2 {
+		return false
+	}
+	minor, err := strconv.ParseInt(parts[1], 10, 32)
+	return err == nil && parts[0] == "1" && minor < 9
+}()
+
+func getCallExprArgs(node ast.Node) ([]ast.Expr, error) {
+	visitor := &callExprVisitor{}
+	ast.Walk(visitor, node)
+	if visitor.expr == nil {
+		return nil, errors.New("failed to find call expression")
+	}
+	debug("usaing (%T): %s", visitor.expr, debugFormatNode{visitor.expr})
+	return visitor.expr.Args, nil
+}
+
+type callExprVisitor struct {
+	expr *ast.CallExpr
+}
+
+func (v *callExprVisitor) Visit(node ast.Node) ast.Visitor {
+	if v.expr != nil || node == nil {
+		return nil
+	}
+	debug("visit (%T): %s", node, debugFormatNode{node})
+
+	switch typed := node.(type) {
+	case *ast.CallExpr:
+		v.expr = typed
+		return nil
+	case *ast.DeferStmt:
+		ast.Walk(v, typed.Call.Fun)
+		return nil
+	}
+	return v
+}
+
+// FormatNode using go/format.Node and return the result as a string
+func FormatNode(node ast.Node) (string, error) {
+	buf := new(bytes.Buffer)
+	err := format.Node(buf, token.NewFileSet(), node)
+	return buf.String(), err
 }
 
 var debugEnabled = os.Getenv("GOTESTTOOLS_DEBUG") != ""

--- a/internal/source/source.go
+++ b/internal/source/source.go
@@ -25,7 +25,7 @@ func FormattedCallExprArg(stackIndex int, argPos int) (string, error) {
 		return "", err
 	}
 	if argPos >= len(args) {
-		return "", errors.New("no arg")
+		return "", errors.New("failed to find expression")
 	}
 	return FormatNode(args[argPos])
 }

--- a/internal/source/source.go
+++ b/internal/source/source.go
@@ -43,7 +43,7 @@ func CallExprArgs(stackIndex int) ([]ast.Expr, error) {
 	if err != nil {
 		return nil, err
 	}
-	debug("found node (%T): %s", node, debugFormatNode{node})
+	debug("found node: %s", debugFormatNode{node})
 
 	return getCallExprArgs(node)
 }
@@ -55,21 +55,26 @@ func getNodeAtLine(filename string, lineNum int) (ast.Node, error) {
 		return nil, errors.Wrapf(err, "failed to parse source file: %s", filename)
 	}
 
-	node := scanToLine(fileset, astFile, lineNum)
-	if node == nil {
-		return nil, errors.Errorf(
-			"failed to find an expression on line %d in %s", lineNum, filename)
+	if node := scanToLine(fileset, astFile, lineNum); node != nil {
+		return node, nil
 	}
-	return node, nil
+	if node := scanToDeferLine(fileset, astFile, lineNum); node != nil {
+		node, err := guessDefer(node)
+		if err != nil || node != nil {
+			return node, err
+		}
+	}
+	return nil, errors.Errorf(
+		"failed to find an expression on line %d in %s", lineNum, filename)
 }
 
 func scanToLine(fileset *token.FileSet, node ast.Node, lineNum int) ast.Node {
 	var matchedNode ast.Node
 	ast.Inspect(node, func(node ast.Node) bool {
-		if node == nil || matchedNode != nil {
+		switch {
+		case node == nil || matchedNode != nil:
 			return false
-		}
-		if nodePosition(fileset, node).Line == lineNum {
+		case nodePosition(fileset, node).Line == lineNum:
 			matchedNode = node
 			return false
 		}
@@ -108,7 +113,7 @@ func getCallExprArgs(node ast.Node) ([]ast.Expr, error) {
 	if visitor.expr == nil {
 		return nil, errors.New("failed to find call expression")
 	}
-	debug("usaing (%T): %s", visitor.expr, debugFormatNode{visitor.expr})
+	debug("callExpr: %s", debugFormatNode{visitor.expr})
 	return visitor.expr.Args, nil
 }
 
@@ -120,7 +125,7 @@ func (v *callExprVisitor) Visit(node ast.Node) ast.Visitor {
 	if v.expr != nil || node == nil {
 		return nil
 	}
-	debug("visit (%T): %s", node, debugFormatNode{node})
+	debug("visit: %s", debugFormatNode{node})
 
 	switch typed := node.(type) {
 	case *ast.CallExpr:
@@ -157,5 +162,5 @@ func (n debugFormatNode) String() string {
 	if err != nil {
 		return fmt.Sprintf("failed to format %s: %s", n.Node, err)
 	}
-	return out
+	return fmt.Sprintf("(%T) %s", n.Node, out)
 }

--- a/internal/source/source_test.go
+++ b/internal/source/source_test.go
@@ -4,19 +4,20 @@ package source_test
 // package
 
 import (
+	"fmt"
 	"testing"
 
 	"gotest.tools/assert"
 	"gotest.tools/internal/source"
 )
 
-func TestGetConditionSingleLine(t *testing.T) {
+func TestFormattedCallExprArg_SingleLine(t *testing.T) {
 	msg, err := shim("not", "this", "this text")
 	assert.NilError(t, err)
 	assert.Equal(t, `"this text"`, msg)
 }
 
-func TestGetConditionMultiLine(t *testing.T) {
+func TestFormattedCallExprArg_MultiLine(t *testing.T) {
 	msg, err := shim(
 		"first",
 		"second",
@@ -26,7 +27,7 @@ func TestGetConditionMultiLine(t *testing.T) {
 	assert.Equal(t, `"this text"`, msg)
 }
 
-func TestGetConditionIfStatement(t *testing.T) {
+func TestFormattedCallExprArg_IfStatement(t *testing.T) {
 	if msg, err := shim(
 		"first",
 		"second",
@@ -39,4 +40,38 @@ func TestGetConditionIfStatement(t *testing.T) {
 
 func shim(_, _, _ string) (string, error) {
 	return source.FormattedCallExprArg(1, 2)
+}
+
+func TestFormattedCallExprArg_InDefer(t *testing.T) {
+	t.Skip("defer reports end of function block as line number")
+	cap := &capture{}
+	func() {
+		fmt.Println()
+		defer fmt.Println()
+		defer cap.shim("first", "second\n")
+	}()
+
+	assert.NilError(t, cap.err)
+	assert.Equal(t, cap.value, `"second"`)
+}
+
+type capture struct {
+	value string
+	err   error
+}
+
+func (c *capture) shim(_, _ string) {
+	c.value, c.err = source.FormattedCallExprArg(1, 1)
+}
+
+func TestFormattedCallExprArg_InAnonymousDefer(t *testing.T) {
+	cap := &capture{}
+	func() {
+		fmt.Println()
+		defer fmt.Println()
+		defer func() { cap.shim("first", "second") }()
+	}()
+
+	assert.NilError(t, cap.err)
+	assert.Equal(t, cap.value, `"second"`)
 }

--- a/internal/source/source_test.go
+++ b/internal/source/source_test.go
@@ -45,7 +45,6 @@ func shim(_, _, _ string) (string, error) {
 func TestFormattedCallExprArg_InDefer(t *testing.T) {
 	cap := &capture{}
 	func() {
-		fmt.Println()
 		defer cap.shim("first", "second")
 	}()
 
@@ -72,4 +71,15 @@ func TestFormattedCallExprArg_InAnonymousDefer(t *testing.T) {
 
 	assert.NilError(t, cap.err)
 	assert.Equal(t, cap.value, `"second"`)
+}
+
+func TestFormattedCallExprArg_InDeferMultipleDefers(t *testing.T) {
+	cap := &capture{}
+	func() {
+		fmt.Println()
+		defer fmt.Println()
+		defer cap.shim("first", "second")
+	}()
+
+	assert.ErrorContains(t, cap.err, "ambiguous call expression")
 }

--- a/internal/source/source_test.go
+++ b/internal/source/source_test.go
@@ -43,12 +43,10 @@ func shim(_, _, _ string) (string, error) {
 }
 
 func TestFormattedCallExprArg_InDefer(t *testing.T) {
-	t.Skip("defer reports end of function block as line number")
 	cap := &capture{}
 	func() {
 		fmt.Println()
-		defer fmt.Println()
-		defer cap.shim("first", "second\n")
+		defer cap.shim("first", "second")
 	}()
 
 	assert.NilError(t, cap.err)

--- a/internal/source/source_test.go
+++ b/internal/source/source_test.go
@@ -5,10 +5,13 @@ package source_test
 
 import (
 	"fmt"
+	"runtime"
+	"strings"
 	"testing"
 
 	"gotest.tools/assert"
 	"gotest.tools/internal/source"
+	"gotest.tools/skip"
 )
 
 func TestFormattedCallExprArg_SingleLine(t *testing.T) {
@@ -43,6 +46,7 @@ func shim(_, _, _ string) (string, error) {
 }
 
 func TestFormattedCallExprArg_InDefer(t *testing.T) {
+	skip.If(t, isGoVersion18)
 	cap := &capture{}
 	func() {
 		defer cap.shim("first", "second")
@@ -50,6 +54,10 @@ func TestFormattedCallExprArg_InDefer(t *testing.T) {
 
 	assert.NilError(t, cap.err)
 	assert.Equal(t, cap.value, `"second"`)
+}
+
+func isGoVersion18() bool {
+	return strings.HasPrefix(runtime.Version(), "go1.8.")
 }
 
 type capture struct {
@@ -74,6 +82,7 @@ func TestFormattedCallExprArg_InAnonymousDefer(t *testing.T) {
 }
 
 func TestFormattedCallExprArg_InDeferMultipleDefers(t *testing.T) {
+	skip.If(t, isGoVersion18)
 	cap := &capture{}
 	func() {
 		fmt.Println()

